File: python/worker/src/transflow_worker/canonical.py
"""Canonical JSON and purpose-separated SHA-256, profile v1.

Input is decoded JSON; duplicate keys must be rejected at the decode boundary.
No source selection, environment inspection, filesystem traversal or publication
occurs here. See schemas/canonical-v1.md for the byte-level contract.
"""

import hashlib
import json
import math
from dataclasses import dataclass
from enum import StrEnum
from typing import BinaryIO

from transflow_worker.wire import ProtocolError, validate_document

MAX_CANONICAL_BYTES = 16 * 1024 * 1024
MAX_SAFE_INTEGER = 9_007_199_254_740_991


class CanonicalError(ValueError):
    """Content violates the canonical profile or a selected fingerprint contract."""


class DigestKind(StrEnum):
    """The role is retained with each digest; FILE hashes raw bytes without a prefix."""

    ARTIFACT = "transflow.artifact.v1"
    SOURCE = "transflow.source.v1"
    COMPUTE = "transflow.compute.v1"
    CATALOG = "transflow.catalog.v1"
    SCHEMA = "transflow.schema.v1"
    FILE = ""


@dataclass(frozen=True)
class ContentDigest:
    """Purpose and lowercase hex content identity, never a publication UUID."""

    kind: DigestKind
    hex: str

    def __post_init__(self) -> None:
        if (
            not isinstance(self.kind, DigestKind)
            or len(self.hex) != 64
            or any(ch not in "0123456789abcdef" for ch in self.hex)
        ):
            raise CanonicalError("Invalid digest role or representation")


def canonical_json(value: object) -> bytes:
    """Serialize with bounded output/depth, exact strings and safe integer metadata."""
    output = bytearray()

    def append(data: bytes) -> None:
        if len(data) > MAX_CANONICAL_BYTES - len(output):
            raise CanonicalError("Canonical metadata exceeds its byte limit")
        output.extend(data)

    def string(text: str) -> None:
        if len(text) > MAX_CANONICAL_BYTES:
            raise CanonicalError("Canonical string exceeds its byte limit")
        append(b'"')
        # Bound temporary escaping allocations even for a large caller-supplied string.
        for offset in range(0, len(text), 256):
            try:
                append(json.dumps(text[offset : offset + 256], ensure_ascii=False)[1:-1].encode())
            except UnicodeEncodeError as error:
                raise CanonicalError("Canonical strings require Unicode scalars") from error
        append(b'"')

    def encode(item: object, depth: int) -> None:
        if depth > 64:
            raise CanonicalError("Canonical metadata exceeds its nesting limit")
        if item is None:
            append(b"null")
        elif isinstance(item, bool):
            append(b"true" if item else b"false")
        elif isinstance(item, int | float):
            if isinstance(item, float) and (not math.isfinite(item) or not item.is_integer()):
                raise CanonicalError("Bare numbers must be finite safe integers")
            if abs(item) > MAX_SAFE_INTEGER:
                raise CanonicalError("Use a lossless string carrier for large numbers")
            append(str(int(item)).encode("ascii"))
        elif isinstance(item, str):
            string(item)
        elif isinstance(item, list):
            append(b"[")
            for index, child in enumerate(item):
                if index:
                    append(b",")
                encode(child, depth + 1)
            append(b"]")
        elif isinstance(item, dict):
            if any(not isinstance(key, str) for key in item):
                raise CanonicalError("JSON object keys must be strings")
            append(b"{")
            for index, key in enumerate(sorted(item)):
                if index:
                    append(b",")
                string(key)
                append(b":")
                encode(item[key], depth + 1)
            append(b"}")
        else:
            raise CanonicalError("Only decoded JSON values can be canonicalized")

    encode(value, 0)
    return bytes(output)


def content_digest(kind: DigestKind, value: object) -> ContentDigest:
    """Hash prefix + NUL + canonical bytes; domain field selection is the caller's job."""
    if not isinstance(kind, DigestKind) or kind is DigestKind.FILE:
        raise CanonicalError("Use file_digest for unprefixed raw file bytes")
    digest = hashlib.sha256(kind.value.encode("ascii") + b"\0")
    digest.update(canonical_json(value))
    return ContentDigest(kind, digest.hexdigest())


def file_digest(reader: BinaryIO) -> ContentDigest:
    """Read every byte in 64 KiB chunks. I/O errors propagate, never yielding a partial hash."""
    digest = hashlib.sha256()
    while True:
        try:
            chunk = reader.read(64 * 1024)
        except InterruptedError:
            continue
        if not isinstance(chunk, bytes):
            raise CanonicalError("Raw content hashing requires a blocking byte stream")
        if not chunk:
            break
        if len(chunk) > 64 * 1024:
            raise CanonicalError("Reader exceeded the requested chunk size")
        digest.update(chunk)
    return ContentDigest(DigestKind.FILE, digest.hexdigest())


def _validate(name: str, value: object) -> None:
    try:
        validate_document(name, value)
    except ProtocolError as error:
        raise CanonicalError("Content does not match its fingerprint contract") from error


def schema_digest(value: object) -> ContentDigest:
    """Validate and fingerprint an ordered logical schema."""
    _validate("LogicalSchemaV1", value)
    return content_digest(DigestKind.SCHEMA, value)


def artifact_digest(value: object) -> ContentDigest:
    """Validate a manifest including its schema fingerprint; file bytes are checked by storage."""
    _validate("ArtifactManifestV1", value)
    if (
        not isinstance(value, dict)
        or value["schema_fingerprint"] != schema_digest(value["logical_schema"]).hex
    ):
        raise CanonicalError("Artifact schema fingerprint does not match")
    return content_digest(DigestKind.ARTIFACT, value)


def catalog_fingerprint(snapshot: object) -> ContentDigest:
    """Exclude the self-fingerprint/source snapshot ID; preserve declared entry order."""
    _validate("CatalogSnapshotV1", snapshot)
    if not isinstance(snapshot, dict):
        raise CanonicalError("A catalogue snapshot must be an object")
    return content_digest(
        DigestKind.CATALOG,
        {key: snapshot[key] for key in ("format_version", "workspace_id", "entries")},
    )


def verify_catalog_fingerprint(snapshot: object) -> None:
    """Reject an otherwise valid snapshot with a stale or tampered fingerprint."""
    digest = catalog_fingerprint(snapshot)
    if not isinstance(snapshot, dict) or snapshot["catalog_fingerprint"] != digest.hex:
        raise CanonicalError("Catalogue fingerprint does not match its content")


def semantic_fingerprint(kind: DigestKind, document: object) -> ContentDigest:
    """Hash only semantic in the closed {semantic, presentation} envelope.

    This explicit split never strips keys named description from user parameters.
    Full source/compute field selection is owned by later capture/planner services.
    """
    if kind not in (DigestKind.SOURCE, DigestKind.COMPUTE) or not isinstance(document, dict):
        raise CanonicalError("Expected a source or compute semantic envelope")
    if set(document) != {"semantic", "presentation"}:
        raise CanonicalError("Semantic envelopes contain exactly semantic and presentation")
    return content_digest(kind, document["semantic"])

File: crates/tf-protocol/src/canonical.rs
//! Canonical JSON and domain-separated SHA-256, profile v1. See schemas/canonical-v1.md.
use crate::validate_document;
use serde_json::Value;
use sha2::{Digest as _, Sha256};
use std::{
    fmt,
    io::{self, Read},
};

/// Maximum serialized metadata size. File bytes are hashed separately in fixed chunks.
pub const MAX_CANONICAL_BYTES: usize = 16 * 1024 * 1024;
/// Largest interoperable bare integer. Larger values use lossless string carriers.
pub const MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;

/// Recoverable canonicalization, contract or file-reading failure.
#[derive(Debug)]
pub enum CanonicalError {
    /// Unsupported number, document, or depth/byte limit.
    Invalid,
    /// Closed contract or fingerprint validation failed.
    Contract,
    /// Raw file hashing failed. A partial digest is never returned.
    Io(io::Error),
}
impl fmt::Display for CanonicalError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Invalid => "Content cannot be encoded with canonical JSON v1",
            Self::Contract => "Content does not match its fingerprint contract",
            Self::Io(_) => "Could not read all content for its digest",
        })
    }
}
impl std::error::Error for CanonicalError {}

/// Hash purposes are retained alongside bytes to prevent accidental role equality.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DigestKind {
    /// Canonical physical artifact manifest.
    Artifact,
    /// Captured source content descriptor (selection belongs to the capture service).
    Source,
    /// Fully selected computation semantics (selection belongs to the planner).
    Compute,
    /// Stable catalogue identity/path manifest.
    Catalog,
    /// Logical schema, including field order.
    Schema,
    /// Unprefixed SHA-256 of raw file bytes, for manifest file entries.
    File,
}
impl DigestKind {
    /// Exact ASCII domain/version prefix; raw file SHA-256 has no prefix.
    pub fn prefix(self) -> &'static str {
        match self {
            Self::Artifact => "transflow.artifact.v1",
            Self::Source => "transflow.source.v1",
            Self::Compute => "transflow.compute.v1",
            Self::Catalog => "transflow.catalog.v1",
            Self::Schema => "transflow.schema.v1",
            Self::File => "",
        }
    }
}
/// Content identity with an explicit role, never a publication/version UUID.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ContentDigest {
    kind: DigestKind,
    bytes: [u8; 32],
}
impl ContentDigest {
    /// Parse a canonical lowercase SHA-256, with the role supplied by its enclosing contract.
    pub fn from_hex(kind: DigestKind, text: &str) -> Result<Self, CanonicalError> {
        if text.len() != 64
            || !text
                .bytes()
                .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
        {
            return Err(CanonicalError::Invalid);
        }
        let mut bytes = [0; 32];
        for (out, pair) in bytes.iter_mut().zip(text.as_bytes().as_chunks::<2>().0) {
            let pair = std::str::from_utf8(pair).map_err(|_| CanonicalError::Invalid)?;
            *out = u8::from_str_radix(pair, 16).map_err(|_| CanonicalError::Invalid)?;
        }
        Ok(Self { kind, bytes })
    }

    /// Purpose used to compute this digest.
    pub fn kind(&self) -> DigestKind {
        self.kind
    }
    /// Lowercase hexadecimal wire representation. Role comes from the enclosing field.
    pub fn hex(&self) -> String {
        self.bytes
            .iter()
            .map(|byte| format!("{byte:02x}"))
            .collect()
    }
}
struct Encoder(Vec<u8>);
impl Encoder {
    fn append(&mut self, bytes: &[u8]) -> Result<(), CanonicalError> {
        if bytes.len() > MAX_CANONICAL_BYTES.saturating_sub(self.0.len()) {
            return Err(CanonicalError::Invalid);
        }
        self.0.extend_from_slice(bytes);
        Ok(())
    }
    fn string(&mut self, s: &str) -> Result<(), CanonicalError> {
        self.append(b"\"")?;
        for ch in s.chars() {
            match ch {
                '"' => self.append(b"\\\"")?,
                '\\' => self.append(b"\\\\")?,
                '\u{8}' => self.append(b"\\b")?,
                '\t' => self.append(b"\\t")?,
                '\n' => self.append(b"\\n")?,
                '\u{c}' => self.append(b"\\f")?,
                '\r' => self.append(b"\\r")?,
                '\u{0}'..='\u{1f}' => self.append(format!("\\u{:04x}", ch as u32).as_bytes())?,
                _ => self.append(ch.encode_utf8(&mut [0; 4]).as_bytes())?,
            }
        }
        self.append(b"\"")
    }
    fn value(&mut self, value: &Value, depth: usize) -> Result<(), CanonicalError> {
        if depth > 64 {
            return Err(CanonicalError::Invalid);
        }
        match value {
            Value::Null => self.append(b"null"),
            Value::Bool(true) => self.append(b"true"),
            Value::Bool(false) => self.append(b"false"),
            Value::String(s) => self.string(s),
            Value::Number(n) => {
                let number = n.as_f64().ok_or(CanonicalError::Invalid)?;
                if !number.is_finite()
                    || number.fract() != 0.0
                    || number.abs() > MAX_SAFE_INTEGER as f64
                {
                    return Err(CanonicalError::Invalid);
                }
                self.append((number as i64).to_string().as_bytes())
            }
            Value::Array(values) => {
                self.append(b"[")?;
                for (index, item) in values.iter().enumerate() {
                    if index != 0 {
                        self.append(b",")?;
                    }
                    self.value(item, depth + 1)?;
                }
                self.append(b"]")
            }
            Value::Object(values) => {
                self.append(b"{")?;
                // Explicit sort remains correct if serde_json's preserve_order feature changes.
                let mut keys: Vec<_> = values.keys().collect();
                keys.sort();
                for (index, key) in keys.into_iter().enumerate() {
                    if index != 0 {
                        self.append(b",")?;
                    }
                    self.string(key)?;
                    self.append(b":")?;
                    self.value(values.get(key).ok_or(CanonicalError::Invalid)?, depth + 1)?;
                }
                self.append(b"}")
            }
        }
    }
}
/// Serialize validated decoded JSON. Reject duplicate keys at the raw decode boundary.
/// Strings are not normalized; wire numeric string spellings are preserved exactly.
pub fn canonical_json(value: &Value) -> Result<Vec<u8>, CanonicalError> {
    let mut encoder = Encoder(Vec::new());
    encoder.value(value, 0)?;
    Ok(encoder.0)
}
/// Hash `ASCII prefix || NUL || canonical JSON`. File digests require `file_digest`.
/// This low-level routine does not select semantic fields or validate a domain schema.
pub fn content_digest(kind: DigestKind, value: &Value) -> Result<ContentDigest, CanonicalError> {
    if kind == DigestKind::File {
        return Err(CanonicalError::Invalid);
    }
    let bytes = canonical_json(value)?;
    let mut hash = Sha256::new();
    hash.update(kind.prefix().as_bytes());
    hash.update([0]);
    hash.update(bytes);
    Ok(ContentDigest {
        kind,
        bytes: hash.finalize().into(),
    })
}
/// SHA-256 over every raw byte, in bounded chunks. Call outside async/write transactions.
pub fn file_digest<R: Read>(reader: &mut R) -> Result<ContentDigest, CanonicalError> {
    let mut hash = Sha256::new();
    let mut buffer = [0; 64 * 1024];
    loop {
        match reader.read(&mut buffer) {
            Ok(0) => break,
            Ok(size) => hash.update(buffer.get(..size).ok_or(CanonicalError::Invalid)?),
            Err(error) if error.kind() == io::ErrorKind::Interrupted => continue,
            Err(error) => return Err(CanonicalError::Io(error)),
        }
    }
    Ok(ContentDigest {
        kind: DigestKind::File,
        bytes: hash.finalize().into(),
    })
}
/// Validate and hash an ordered logical schema.
pub fn schema_digest(value: &Value) -> Result<ContentDigest, CanonicalError> {
    validate_document("LogicalSchemaV1", value).map_err(|_| CanonicalError::Contract)?;
    content_digest(DigestKind::Schema, value)
}
/// Validate a physical manifest and its schema fingerprint before hashing all fields.
/// File digests/lengths are asserted metadata here; storage verifies actual files later.
pub fn artifact_digest(value: &Value) -> Result<ContentDigest, CanonicalError> {
    validate_document("ArtifactManifestV1", value).map_err(|_| CanonicalError::Contract)?;
    if value["schema_fingerprint"].as_str()
        != Some(schema_digest(&value["logical_schema"])?.hex().as_str())
    {
        return Err(CanonicalError::Contract);
    }
    content_digest(DigestKind::Artifact, value)
}
/// Select catalogue content, excluding its self-fingerprint and independent source snapshot ID.
/// Entry array order is significant; producers must supply a deterministic order.
pub fn catalog_fingerprint(snapshot: &Value) -> Result<ContentDigest, CanonicalError> {
    validate_document("CatalogSnapshotV1", snapshot).map_err(|_| CanonicalError::Contract)?;
    content_digest(
        DigestKind::Catalog,
        &serde_json::json!({
            "format_version": snapshot["format_version"],
            "workspace_id": snapshot["workspace_id"],
            "entries": snapshot["entries"],
        }),
    )
}
/// Reject a catalogue snapshot whose recorded fingerprint does not describe its content.
pub fn verify_catalog_fingerprint(snapshot: &Value) -> Result<(), CanonicalError> {
    if snapshot["catalog_fingerprint"].as_str()
        != Some(catalog_fingerprint(snapshot)?.hex().as_str())
    {
        return Err(CanonicalError::Contract);
    }
    Ok(())
}
/// Hash an explicitly selected semantic document, never recursively stripping user keys.
/// The closed envelope has exactly `semantic` and `presentation`; presentation is excluded.
/// Only source/compute use this envelope. Complete source/compute field selection is later work.
pub fn semantic_fingerprint(
    kind: DigestKind,
    document: &Value,
) -> Result<ContentDigest, CanonicalError> {
    if !matches!(kind, DigestKind::Source | DigestKind::Compute) {
        return Err(CanonicalError::Contract);
    }
    let members = document.as_object().ok_or(CanonicalError::Contract)?;
    if members.len() != 2 || !members.contains_key("presentation") {
        return Err(CanonicalError::Contract);
    }
    content_digest(
        kind,
        members.get("semantic").ok_or(CanonicalError::Contract)?,
    )
}
